xhype/xhype/src/vmexit.rs

```diff
@@ -13,6 +16,105 @@ pub enum HandleResult {
     Next,
 }
 
+// Table 27-3
+#[inline]
+pub fn vmx_guest_reg(num: u64) -> X86Reg {
+    match num {
+        0 => X86Reg::RAX,
+        1 => X86Reg::RCX,
+        2 => X86Reg::RDX,
+        3 => X86Reg::RBX,
+        4 => X86Reg::RSP,
+        5 => X86Reg::RBP,
+        6 => X86Reg::RSI,
+        7 => X86Reg::RDI,
+        8 => X86Reg::R8,
+        9 => X86Reg::R9,
+        10 => X86Reg::R10,
+        11 => X86Reg::R11,
+        12 => X86Reg::R12,
+        13 => X86Reg::R13,
+        14 => X86Reg::R14,
+        15 => X86Reg::R15,
+        _ => panic!("bad register in exit qualification"),
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// VMX_REASON_MOV_CR
+////////////////////////////////////////////////////////////////////////////////
+
+// Intel SDM Table 27-3.
+pub fn handle_cr(vcpu: &VCPU, _gth: &GuestThread) -> Result<HandleResult, Error> {
+    let qual = vcpu.read_vmcs(VMCS_RO_EXIT_QUALIFIC)?;
```

Comment:
Qualification is sufficiently complex here that you may consider wrapping it up inside of e.g. a `bitfield` (https://crates.io/crates/bitfield):

```
bitfield! {
    #[derive(Copy, Clone)]
    pub struct MovCr0Qual(u32);
    impl Debug;
    raw_dest, _: 3, 0;
    pub access_type, _: 5, 4;
    raw_src, _: 11, 8;
}

type Error = &'static str;
type Result<T> = result::Result<T, Error>;
impl MovCr0Qual {
    pub fn ctrl_reg(self) -> Result<CtrlReg> {
        match self.raw_dest() {
            0 => Ok(X86Reg::CR0),
            3 => Ok(X86Reg::CR3),
            4 => Ok(X86Reg::CR4),
            _ => Err("unsupported control register"),
        }
    }

    pub fn src_reg(self) -> Result<Regs> {
        Regs::try_from(self.raw_src())
    }
}
```

Alternatively, instead of doing `try_from()` in `src_reg()` you could just do the match there.

and you write your code in terms of those things (`match qual.ctrl_reg()` and so on).